subsys/net/lib/websocket/websocket.c

```diff
@@ -715,94 +724,130 @@ int websocket_send_msg(int ws_sock, const uint8_t *payload, size_t payload_len,
-				   uint32_t *message_type_flag,
-				   size_t *header_len)
-{
-	uint8_t len_len; /* length of the length field in header */
-	uint8_t len;     /* message length byte */
-	uint16_t value;
 
-	value = sys_get_be16(&buf[0]);
-	if (value & 0x8000) {
-		*message_type_flag |= WEBSOCKET_FLAG_FINAL;
+static uint32_t websocket_opcode2flag(uint8_t data)
+{
+	switch (data & 0x0f) {
+	case WEBSOCKET_OPCODE_DATA_TEXT:
+		return WEBSOCKET_FLAG_TEXT;
+	case WEBSOCKET_OPCODE_DATA_BINARY:
+		return WEBSOCKET_FLAG_BINARY;
+	case WEBSOCKET_OPCODE_CLOSE:
+		return WEBSOCKET_FLAG_CLOSE;
+	case WEBSOCKET_OPCODE_PING:
+		return WEBSOCKET_FLAG_PING;
+	case WEBSOCKET_OPCODE_PONG:
+		return WEBSOCKET_FLAG_PONG;
+	default:
+		break;
 	}
+	return 0;
+}
 
-	switch (value & 0x0f00) {
-	case 0x0100:
-		*message_type_flag |= WEBSOCKET_FLAG_TEXT;
+static int websocket_parser_step(struct websocket_context *ctx, uint8_t data)
+{
+	int len;
+
+	switch (ctx->parser_state) {
+	case WEBSOCKET_PARSER_STATE_OPCODE:
+		ctx->message_type = websocket_opcode2flag(data);
+		if ((data & 0x80) != 0) {
+			ctx->message_type |= WEBSOCKET_FLAG_FINAL;
+		}
+		ctx->parser_state = WEBSOCKET_PARSER_STATE_LENGTH;
 		break;
-	case 0x0200:
-		*message_type_flag |= WEBSOCKET_FLAG_BINARY;
+	case WEBSOCKET_PARSER_STATE_LENGTH:
+		ctx->masked = (data & 0x80) != 0;
+		len = data & 0x7f;
+		if (len < 126) {
+			ctx->message_len = len;
+			if (ctx->masked) {
+				ctx->masking_value = 0;
+				ctx->parser_remaining = 4;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_MASK;
+			} else {
+				ctx->parser_remaining = ctx->message_len;
+				ctx->parser_state = (ctx->parser_remaining == 0)
+							? WEBSOCKET_PARSER_STATE_OPCODE
+							: WEBSOCKET_PARSER_STATE_PAYLOAD;
+			}
+		} else {
+			ctx->message_len = 0;
+			ctx->parser_remaining = (len < 127) ? 2 : 8;
+			ctx->parser_state = WEBSOCKET_PARSER_STATE_EXT_LEN;
+		}
 		break;
-	case 0x0800:
-		*message_type_flag |= WEBSOCKET_FLAG_CLOSE;
+	case WEBSOCKET_PARSER_STATE_EXT_LEN:
+		ctx->parser_remaining--;
+		ctx->message_len |= (data << (ctx->parser_remaining * 8));
+		if (ctx->parser_remaining == 0) {
+			if (ctx->masked) {
+				ctx->masking_value = 0;
+				ctx->parser_remaining = 4;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_MASK;
+			} else {
+				ctx->parser_remaining = ctx->message_len;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_PAYLOAD;
+			}
+		}
 		break;
-	case 0x0900:
-		*message_type_flag |= WEBSOCKET_FLAG_PING;
+	case WEBSOCKET_PARSER_STATE_MASK:
+		ctx->parser_remaining--;
+		ctx->masking_value |= (data << (ctx->parser_remaining * 8));
+		if (ctx->parser_remaining == 0) {
+			if (ctx->message_len == 0) {
+				ctx->parser_remaining = 0;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_OPCODE;
+			} else {
+				ctx->parser_remaining = ctx->message_len;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_PAYLOAD;
+			}
+		}
 		break;
-	case 0x0A00:
-		*message_type_flag |= WEBSOCKET_FLAG_PONG;
+	case WEBSOCKET_PARSER_STATE_PAYLOAD:
```

Comment:
Just a general note for discussion - I wonder if the payload processing could be optimized a bit, i. e. use memcpy instead of processing byte by byte.

Reply:
At first i thought it will be difficult, but i made this change.